Sources/ApodiniDeployRuntimeSupport/StructureExporter.swift

```diff
@@ -53,12 +53,33 @@ extension StructureExporter {
     ) throws -> AnyDeployedSystem {
         try retrieveDefaultDeployedSystem(endpoints, config: config, app: app)
     }
-    
+
     /// Computes the default deployed system where all endpoints are matched to deployment groups if possible.
-    public func retrieveDefaultDeployedSystem(
+    public  func retrieveDefaultDeployedSystem(
         _ endpoints: Set<CollectedEndpointInfo>,
         config: DeploymentConfig,
         app: Application
+    ) throws -> DeployedSystem {
+        try retrieveDefaultDeployedSystem(endpoints, config: config, app: app, deriveUserData: nil)
+    }
+
+    /// Computes the default deployed system where all endpoints are matched to deployment groups if possible.
+    public func retrieveDefaultDeployedSystem<T: Codable>(
+        _ endpoints: Set<CollectedEndpointInfo>,
+        config: DeploymentConfig,
+        app: Application,
+        deriveUserData: @escaping (CollectedEndpointInfo) -> T
+    ) throws -> DeployedSystem {
```

Comment:
Maybe I'm missing something, but was exactly is the purpose of this method and in particular the `deriveUserData` parameter? I'd assume it's to manually decide how to handle/ decode the user data in a custom type, similar to what was talked in rocket chat, is this correct?  
But this parameter does only work for whoever how wants to use the default deployed system, or am i getting this wrong? Anybody relying on a custom implementation cannot use this. 
Would it be possible to add this to the `StructureExporter` protocol with a default `nil` implementation similar to `nodeIdProvider`, then everybody implementing this protocol could decide if he wants to use it or not

Reply:
The `retrieveDefaultDeployedSystem` of the `StructureExporter` creates `ExportedEndpoint`s from `CollectedEndpointInfo`s.
`CollectedEndpointInfo` contains deployment options, the `ExportedEndpoint` contains a `Data` property which, similar to `DeployedSystemNode`, is en and decoded from a `Codable` type provided by the deployment provider.
This basically realizes the "context" thingy we talked about.
In order, to not be required to write your own `retrieveDeployedSystem` if you want to interpret deployment options and store them in your `userData`, we provide the closure which will then create such a userData instance from the provided `CollectedEndpointInfo`.

Reply:
And yes, good catch. Completely forgot to add it to the protocol itself 👍

Reply:
> Would it be possible to add this to the StructureExporter protocol with a default nil implementation similar to nodeIdProvider, then everybody implementing this protocol could decide if he wants to use it or not

> And yes, good catch. Completely forgot to add it to the protocol itself 👍

Nevermind. This method is not there to be implemented, it is there to be used in your own `retrieveDeploymentStructure`. Similarly, as both `LambdaStructureExporterCommand` and `LocalhostStructureExporterCommand` call the `retrieveDefaultDeployedSystem` without the closure in their implementation.

Reply:
> Anybody relying on a custom implementation cannot use this.

Anybody doing a custom implementation can just instantiate a `ExportedEndpoint` with according userData themselves.

Reply:
Ah, I see. I guess, i mixed it up with [WellKnownEnvironmentVariables](https://github.com/Apodini/Apodini/blob/a990489b5b392ad34d66a18da14ed3a2335d6c4c/Sources/ApodiniDeployBuildSupport/CollectedEndpointInfo.swift). 

Sure, never mind. I'll just clean them up in a later pr. 👍